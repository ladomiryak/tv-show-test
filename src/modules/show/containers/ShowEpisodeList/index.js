import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import Component from "./ShowEpisodeList";
import * as showActions from "../../actions";
import { showSelectors } from "../../selectors";

const mapStateToProps = (state) => ({
  episodeList: showSelectors.getShowEpisodeList(state),
  show: showSelectors.getShow(state),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      fetchShowEpisodeList: showActions.fetchShowEpisodeList,
    },
    dispatch
  );

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(Component);

export default ConnectedComponent;
