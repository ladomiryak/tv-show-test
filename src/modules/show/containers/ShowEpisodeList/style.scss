@import "../../.././../assets/scss/colors";

.show-list {
  margin-bottom: 24px;
  position: relative;

  // Scroll to see more overlay
  &::after {
    content: "";
    position: absolute;
    right: -8px;
    top: 0;
    bottom: 0;
    width: 70px;
    border-radius: 0 4px 4px 0;
    background-image: linear-gradient(
      to right,
      transparent,
      rgba($primary, 0.1)
    );
  }

  &__season {
    margin-bottom: 16px;
    display: inline-flex;
    font-size: 16px;
    background-color: $tertiary;
    border-radius: 4px;
    padding: 4px 8px;
    color: $white;
  }

  &__items {
    display: flex;
    overflow: auto;
    margin: 0 -8px;
  }

  &__item {
    margin-right: 8px;
    flex: 1 0 220px;
    padding: 8px;
  }

  &__item-poster {
    border-radius: 4px;
    box-shadow: 0 0 10px 0 rgba($primary, 0.1);
    margin-bottom: 8px;
    width: 100%;
    display: block;
    transition: 250ms;
  }

  &__item-link {
    display: block;
    text-decoration: none;

    &:hover {
      img {
        box-shadow: 0 5px 10px 3px rgba($primary, 0.15);
      }
    }
  }

  &__item-title {
    margin-bottom: 8px;
    font-size: 16px;
    display: block;
    color: $primary;
  }
}
