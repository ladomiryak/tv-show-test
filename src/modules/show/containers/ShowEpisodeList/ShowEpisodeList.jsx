import React, { useState, useEffect } from "react";
import { groupBy } from "../../../../utils/common";
import { Loader, Alert } from "../../../shared";
import { Link } from "react-router-dom";
import "./style.scss";

const ShowEpisodeList = ({
  error,
  showId,
  episodeList,
  fetchShowEpisodeList,
  showSeason,
}) => {
  const [listFiltered, setListFiltered] = useState([]);

  // fetch Episode List
  useEffect(() => {
    if (showId && !episodeList) {
      fetchShowEpisodeList({
        showId,
      });
    }
  }, [showId, fetchShowEpisodeList, episodeList]);

  // Group Episode list by season
  useEffect(() => {
    if (episodeList) {
      setListFiltered(groupBy(episodeList, "season"));
    }
  }, [episodeList]);

  // Render Episode Item
  const EpisodeItem = ({ name, season, number, image }) => (
    <li className="show-list__item">
      <Link
        to={`/show/${showId}/season/${season}/episode/${number}`}
        className="show-list__item-link"
      >
        <img
          src={
            image?.medium ||
            "https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
          }
          alt="Episode poster"
          className="show-list__item-poster"
        />

        <h4 className="show-list__item-title">{name}</h4>
      </Link>
    </li>
  );

  if (error) {
    return (
      <Alert variant="error">Something went wrong. Try later, please.</Alert>
    );
  }

  if (!episodeList) {
    return <Loader />;
  }

  return (
    <ul>
      {showSeason && listFiltered[showSeason] ? (
        // Show Selected Season Episode list
        <li className="show-list">
          <span className="show-list__season">Season: {showSeason}</span>
          <ul className="show-list__items">
            {listFiltered[showSeason].map((episode) => (
              <EpisodeItem key={episode.number} {...episode} />
            ))}
          </ul>
        </li>
      ) : (
        // Show full Episode List
        Object.keys(listFiltered).map((key) => (
          <li className="show-list" key={key}>
            <span className="show-list__season">Season: {key}</span>
            <ul className="show-list__items">
              {listFiltered[key].map((episode) => (
                <EpisodeItem key={episode.number} {...episode} />
              ))}
            </ul>
          </li>
        ))
      )}
    </ul>
  );
};

export default ShowEpisodeList;
