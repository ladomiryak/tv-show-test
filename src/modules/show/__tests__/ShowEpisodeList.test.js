import React from "react";
import ShowEpisodeList from "../containers/ShowEpisodeList";
import { render, screen } from "../test-util";
import stateMock from "./mock.json";

describe("Show Episode List", () => {
  test("Loader should be shown with empty state", () => {
    const { container } = render(<ShowEpisodeList />, {
      initialState: { show: {} },
    });

    expect(screen.getByTestId("loader")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("Three seasons should be shown", () => {
    const state = jest.fn(() => stateMock);
    const { container } = render(<ShowEpisodeList />, {
      initialState: {
        show: state(),
      },
    });

    expect(container.querySelectorAll(".show-list__season").length).toEqual(3);
  });

  test("Season full episode list should be show", () => {
    const state = jest.fn(() => stateMock);
    const { container } = render(<ShowEpisodeList />, {
      initialState: {
        show: state(),
      },
    });

    expect(container.querySelectorAll(".show-list__item").length).toEqual(119);
  });

  test("Season third episode list should be show", () => {
    const state = jest.fn(() => stateMock);
    const { container } = render(<ShowEpisodeList showSeason={3} />, {
      initialState: {
        show: state(),
      },
    });

    expect(container.querySelectorAll(".show-list__item").length).toEqual(39);
    expect(container).toMatchSnapshot();
  });
});
