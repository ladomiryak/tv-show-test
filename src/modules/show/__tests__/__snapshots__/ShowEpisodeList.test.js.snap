// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show Episode List Loader should be shown with empty state 1`] = `
<div>
  <div
    class="loader"
    data-testid="loader"
  />
</div>
`;

exports[`Show Episode List Season third episode list should be show 1`] = `
<div>
  <ul>
    <li
      class="show-list"
    >
      <span
        class="show-list__season"
      >
        Season: 
        3
      </span>
      <ul
        class="show-list__items"
      >
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/1"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="http://static.tvmaze.com/uploads/images/medium_landscape/156/390583.jpg"
            />
            <h4
              class="show-list__item-title"
            >
              Not So Secret Service
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/2"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="http://static.tvmaze.com/uploads/images/medium_landscape/157/394148.jpg"
            />
            <h4
              class="show-list__item-title"
            >
              Worship
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/3"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Blossom3
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/4"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Mojo the Great
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/5"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Trouble Clef
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/6"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Save the Date
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/7"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Can't Buy Love
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/8"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Largo
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/9"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Blundercup
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/10"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Ragnarock and Roll
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/11"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Total Eclipse Of The Kart, Parts 1 & 2
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/12"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              The Long Skate Home
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/13"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Quarantine
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/14"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              In The Doghouse
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/15"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Salamander
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/16"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Small World: Abra-disaster
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/17"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Small World: Stone Cold Spider
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/18"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Small World: Maze Daze
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/19"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Small World: Heart To Heartstone
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/20"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Witch's Crew
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/21"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              The Gift
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/22"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Oh, Daisy!
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/23"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Brain Freeze
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/24"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Lights Out!
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/25"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Bucketboy!
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/26"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              The Fog
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/27"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              The Spoon
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/28"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Cat Burglar
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/29"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Hustlecup
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/30"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Rebel Rebel
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/31"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Our Brand is Chaos
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/32"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Tagalong
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/33"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Brainlord
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/34"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Checkin' Out!
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/35"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Drama Bomb
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/36"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Watch It!
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/37"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Man Up 4: The Donnyest Game
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/38"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              The Oct-father
            </h4>
          </a>
        </li>
        <li
          class="show-list__item"
        >
          <a
            class="show-list__item-link"
            href="/show/undefined/season/3/episode/39"
          >
            <img
              alt="Episode poster"
              class="show-list__item-poster"
              src="https://via.placeholder.com/220x120/FFFFFF/000000?Text=Poster"
            />
            <h4
              class="show-list__item-title"
            >
              Sideline Dad
            </h4>
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
